File: Python/veículo.py
#veiculo.py

#herança = É UM

#clase mãe
class Automovel:

    #construtor
    def __init__(self,modelo):
        self.modelo = modelo
        self.combustivel = 0                                     
        self.kmPorLitro = 8.0

    #função biultin
    #retorna uma string
    def __str__(self):
        return 'Automovel modelo: ' + self.modelo
    
    #funções de classe
    def ligar(self):
        self.ligado = True
        print("%s ligado" % self.modelo)

    def desligar(self):
        self.ligado = False
        print("%s desligado" % self.modelo)

    def acelerar(self):
        print("Acelerando %s VRummm..." % self.modelo)

    def ligar_alarme(self):
        print("Ligando alarme do %s BIP BIP" % self.modelo)

    def desligar_alarme(self):
        print("Desligando alarme do %s BIP BIP" % self.modelo)

    def abastecer(self, combustivel_litros):
        self.combustivel = self.combustivel + combustivel_litros
        print("{} abasteceu {} litros".format(self.modelo, combustivel_litros))

    def rodar(self, kmParaRodar):
        if(not self.ligado):
            print("{} está desligado!".format(self.modelo))
        elif ((self.combustivel * self.kmPorLitro) >= kmParaRodar):
            print("{} rodando {} km" .format(self.modelo, kmParaRodar))
            print("Combustível no tanque antes de rodar = {} litros".format (self.combustivel))
            self.combustivel = (self.combustivel - (kmParaRodar / self.kmPorLitro))
            print("Combustível restante após rodar = {} litros".format(self.combustivel))
        else:
            print("Nao ha combustivel suficiente para rodar!")
            
#classe filho (carro é um automovel
class Carro(Automovel):
    #sobrescreve a função str
    def __str__(self):
        return 'Carro modelo: ' + self.modelo

class Moto(Automovel):
    def __str__(self):
        return 'Moto modelo: ' + self.modelo


#main
fusca = Carro('Fusca')
fusca.ligar()
fusca.acelerar()
fusca.ligar_alarme()
fusca.desligar_alarme()
fusca.abastecer(50)
fusca.rodar(50)
fusca.desligar()

moto = Moto('Moto CG')
moto.ligar()


auto = Automovel('???')

print(auto)
print(fusca)
print(moto)

#fusca é um carro?
if (isinstance(fusca, Carro)):
    print ('Fusca é um carro')
else:
    print('Fusca não é um carro')

File: C/veículo.cpp
//veiculo.cpp

#include <iomanip>
#include <iostream>
#include <stdlib.h>
#include <windows.h>

using namespace std;

class Veiculo {
    
    private:
        //membros de dados de classe
        string marca, modelo, placa, cor;
        int ano;
        float preco;
        float combustivel;
        float kmPorLitro; //em litros
        bool ligado;
        
   public:   
        //construtor da classe
        //é uma função especial que é chamada automaticamente quando o objeto é criado
        Veiculo(string modelo, float kmPorLitro){
            this->modelo = modelo;           
            this->kmPorLitro = kmPorLitro;
            this->ligado = false;
    }     
            //funções=verbo  
    void ligar(){
        this->ligado = true;
        cout << modelo << " ligado" << endl;
    }
    
     void desligar(){
        this->ligado = false;
        cout << modelo << " desligado" << endl;
    }
    
    void acelerar(){
        cout << modelo << "acelerando... Vrummm..."<< endl;
    }
    
    void ligarAlarme(int frequencia){
        cout << "\nLigando alarme... \n";
        Beep(frequencia +1000,300);
        Sleep(100);
        Beep(frequencia +1000,300);
    }
    
    void desligarAlarme(int frequencia){
        cout << "\nDesligando alarme... \n";
        Beep(frequencia ,300);
        Sleep(100);
        Beep(frequencia, 300);
    }
    
    void abastecer(float combustivel_litros){
        this->combustivel += combustivel_litros;
        cout << "\n" << modelo << " abasteceu " << combustivel_litros << " litros " << endl;
       
    }
    
     void rodar(float kmParaRodar){
        cout << setprecision(1) << fixed;
        if (!this->ligado){
        cout  << modelo << " desligado! " << endl;
      } else if ((combustivel * kmPorLitro) >= kmParaRodar) {
        cout  << modelo << " rodando... " << kmParaRodar << " km " << endl 
              <<"Combustivel no tanque antes de rodar = "
              << combustivel << " litros. " << endl;
        this->combustivel -= (kmParaRodar / kmPorLitro);
        cout << "Combustivel restante apos rodar = "
             << combustivel << " litros." << endl;
    }else
    cout << "Nao ha combustivel suficiente para rodar!" << endl;
    }
}; //fim da classe 
    
int main(){  
    
    //instancia um objeto veicuulo
    //             chama o construtor
    
   cout << "--------Fusca---------" << endl;
    Veiculo fusca = Veiculo("Fusca", 8.0f);
    fusca.ligar();     
    fusca.acelerar();
    fusca.ligarAlarme(3000);
    fusca.desligarAlarme(3000);
    fusca.abastecer(50);
    fusca.rodar(10000);
    fusca.desligar(); 
    
    cout << "--------Celta---------" << endl;
    Veiculo celta = Veiculo ("Celta",15.0f);
    celta.ligar();    
    celta.acelerar();
    celta.ligarAlarme(2000);
    celta.desligarAlarme(2000);
    celta.abastecer(80);
    celta.rodar(10000);
    celta.desligar(); 
    
    system("pause");
    return 0;
}
